File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int,int>mp;
        int maxlen=0;
        int n=fruits.size();
        int l=0;
        int r=0;
        while(r<n)
        {
            mp[fruits[r]]++;

            if(mp.size()>2)
            {
                mp[fruits[l]]--;

                if(mp[fruits[l]]==0)
                {
                mp.erase(fruits[l]);
                }
                l++;
            }
            if(mp.size()<=2)
            {
                maxlen=max(maxlen,r-l+1);
            }
            r++;

        }
        return maxlen;
       
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {

        int l=0;
        int r=0;
        int n=fruits.length;
        int maxlen=0;
        HashMap<Integer,Integer>mp=new HashMap<>();

        while(r<n)
        {
            mp.put(fruits[r],mp.getOrDefault(fruits[r],0)+1);

            
            while(mp.size()>2)
            {
                mp.put(fruits[l],mp.get(fruits[l])-1);
               
                if(mp.get(fruits[l])==0)
                {
                    mp.remove(fruits[l]);
                }
                l++;
            }
           
            maxlen=Math.max(maxlen,r-l+1);
            r++;
        }
        return maxlen;
        
    }
}